### DISCUSSION TITLE
elasticsearch container started and then exited

### DISCUSSION THREAD

**Discussion by Korrozia** [testcontainers/testcontainers-dotnet#460] [Q&A] [answered]
Hi all,

I'm trying to use elasticsearch for integration test:
```
private TestcontainersContainer? container;
......

[TestInitialize]
public void TestInitialize()
{
container = new TestcontainersBuilder<TestcontainersContainer>()
							.WithEnvironment("-e", "\"discovery.type=single-node\"")
							.WithEnvironment("-p", "127.0.0.1:9200:9200")
							.WithEnvironment("-p", "127.0.0.1:9300:9300")
							.WithImage("docker.elastic.co/elasticsearch/elasticsearch:7.17.3")
							.WithName("docker-elasticsearch-test")
							.Build();

			container.StartAsync().Wait();
}
```

I tried this command: docker run -p 127.0.0.1:9200:9200 -p 127.0.0.1:9300:9300 -e "discovery.type=single-node" docker.elastic.co/elasticsearch/elasticsearch:7.17.3, and it is working correct.

But when I create docker-elasticsearch-test container with [dotnet-testcontainers](https://github.com/HofmeisterAn/dotnet-testcontainers) library then docker-elasticsearch-test container started and after 5-8 seconds exited. Also http://localhost:9200/ also doesn't work.

Found these errors in logs:
ERROR: [2] bootstrap checks failed. You must address the points described in the following [2] lines before starting Elasticsearch.

bootstrap check failure [1] of [2]: max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]

bootstrap check failure [2] of [2]: the default discovery settings are unsuitable for production use; at least one of [discovery.seed_hosts, discovery.seed_providers, cluster.initial_master_nodes] must be configured

These errors most probably because elasticsearch not started in single node, i.e. I suspect .WithEnvironment("-e", "\"discovery.type=single-node\"") doesn't work.

**How can I fixed this?**

**Comment by HofmeisterAn** [comment, marked as answer]
> I tried this command: docker run -p 127.0.0.1:9200:9200 -p 127.0.0.1:9300:9300 -e "discovery.type=single-node" docker.elastic.co/elasticsearch/elasticsearch:7.17.3, and it is working correct.

1. Your builder configuration does not match the `docker run` command.
2. You don't expose and bind the ports (that's necessary to access your service via HTTP).
3. `WithEnvironment` gets the name and value of the env variable like: `WithEnvironment("discovery.type", "single-node")`
4. Probably, you'll need a custom wait strategy to detect when the container is up and running.

#438 explains some best practices. See also this [example](https://github.com/HofmeisterAn/dotnet-testcontainers/issues/421#issuecomment-1050133740) regarding CosmosDB. I also wouldn't recommend synchronization methods like `Wait()`.

**Comment by Korrozia** [comment]
Thank you! Your reply helped me to find the reason of problem, this is a working code:
```
[TestInitialize]
		public void TestInitialize()
		{
			container = new TestcontainersBuilder<TestcontainersContainer>()
							.WithImage("docker.elastic.co/elasticsearch/elasticsearch:7.17.3")
							.WithName("docker-elasticsearch-test")
							.WithEnvironment("discovery.type", "single-node")
							.WithExposedPort(9200)
							.WithExposedPort(9300)
							.WithPortBinding(9200, 9200)
							.WithPortBinding(9300, 9300)
							.WithWaitStrategy(Wait.ForUnixContainer().UntilPortIsAvailable(9200))
							.Build();

			container.StartAsync().Wait();
		}
```

**Reply by HofmeisterAn** [reply]
As recommended in the best practices, you should assign random ports.
